Add basic functionality for multi-series chart values

paulrosenzweig opened the pull request:
The first commit here adds basic labeling. It doesn't yet update how we drop labels to avoid crowding or add series-level settings.
# Examples
## Grouped Bar
![image](https://user-images.githubusercontent.com/691495/83678723-2f9d0900-a5ac-11ea-9ae6-247289f78693.png)
## Multiple Lines
Lots of crowding 😬 
![image](https://user-images.githubusercontent.com/691495/83678757-39bf0780-a5ac-11ea-9785-69c46c803bbf.png)
## Stacked Bar
![image](https://user-images.githubusercontent.com/691495/83678775-417eac00-a5ac-11ea-9d22-a81a0bb6d868.png)
## Stacked Histogram
![image](https://user-images.githubusercontent.com/691495/83678789-49d6e700-a5ac-11ea-839e-0afeb6d9e7f8.png)
## Stacked Area
![image](https://user-images.githubusercontent.com/691495/83678812-52c7b880-a5ac-11ea-869f-a7a0d7eee8d4.png)
## Combo Bar/Line
![image](https://user-images.githubusercontent.com/691495/83678850-61ae6b00-a5ac-11ea-82a2-89647fa10ba9.png)

paulrosenzweig committed:
add basic functionality for multi-series chart values

paulrosenzweig committed:
fix eslint

paulrosenzweig committed:
fix some bugs with non-ordinal x axes

mazameli commented:
I went through the current state of things just to note some specifics I'm seeing so far.

Looks like split y-axes cause problems currently:

![image](https://user-images.githubusercontent.com/2223916/84090715-60fc4700-a9a7-11ea-8e79-ad77ad3402a6.png)

Combo charts weren't working for me if I had two metrics and one dimension:

![image](https://user-images.githubusercontent.com/2223916/84091016-247d1b00-a9a8-11ea-83c4-5e522cbdc2b7.png)

I experienced some weirdness with an unstacked bar chart when I had multiple metrics with a single dimension:

![image](https://user-images.githubusercontent.com/2223916/84091178-8b023900-a9a8-11ea-8dab-820acf3b8a49.png)

paulrosenzweig committed:
fix split y axis

paulrosenzweig committed:
split spacing logic and auto-compact formatting for each series

paulrosenzweig committed:
fix bugs in non-ordinal grouped bars

paulrosenzweig committed:
allow switching off values for individual series

paulrosenzweig commented:
I fixed the split y-axis issue and some of the other bugs. I still don't a great way to avoid overlaps without dropping labels randomly as they overlap. I worry that will look scattered.

I added a setting to hide labels for individual series, which deviates from the checkbox in @kdoh's designs. That will at least let people _manually_ avoid overlap.

![image](https://user-images.githubusercontent.com/691495/84196517-52c52e00-aa6e-11ea-8a20-0fb47aee0c51.png)

Should I also move the compact formatting setting down to each series for overriding?

paulrosenzweig committed:
fix series value hiding for stacked

paulrosenzweig committed:
drop clustered line labels

mazameli commented:
Items from the demo today:
- Put the global settings block above the individual series blocks?
- Should I be able to turn values on for a single series without first having to turn labels on globally? E.g. if I have a metric by time and by State, and I only want to turn values on for a single state that I'm trying to highlight, like CA, I shouldn't have to turn values on globally and then manually turn off 49 other states.
- Explore rotating values 45 or 90 degrees when there are many of them?

paulrosenzweig committed:
move series settings to bottom for line and bar

paulrosenzweig committed:
allow enabling series values with global setting turned off

paulrosenzweig commented:
@mazameli & @kdoh, I pushed changes to put the series settings at the end for line and bar charts (ordering is implicitly defined on each chart 😕) and allow enabling individual series when the global setting is off.

Currently the only way to access spacing and formatting sub-options is with the global setting enabled. If we want to keep this, I guess we'll need to nest those in each series.

---

Also, what do you think about this "sunken" 45º label when grouped bars get too thin? It avoids needing to redo margins to accommodate the additional height.

![image](https://user-images.githubusercontent.com/691495/84449934-c5740c00-ac1c-11ea-989b-7da3c060e49f.png)

mazameli commented:
> Also, what do you think about this "sunken" 45º label when grouped bars get too thin? It avoids needing to redo margins to accommodate the additional height.

I'm afraid this really doesn't work for me. If we're gonna do something like this, the more conventional treatment for bars would be [90º _inside_ the bars](https://www.highcharts.com/demo/column-rotated-labels). However, introducing that style seems like it would beg for a toggle to choose between inside or outside the bar, which sounds like a one-way ticket to Knobtown.

I'm spinning up the branch now to give it all a look and play with it to form some more opinions on what to do with crowded cases.

mazameli commented:
Did the `Data` tab intentionally get moved to the end here? It should be first.

![image](https://user-images.githubusercontent.com/2223916/84549776-bd1fdd80-acbd-11ea-9a05-cbec118d3aca.png)

mazameli commented:
I started a new simple question, summarized by count, date, and product category, then picked Bar chart. The values didn't display by default, but weirdly, in the settings side panel, the toggle was turned on. If I turn it off then on again, the values appear. If I then click the little reset icon, the toggle turns on, but the values disappear.

![image](https://user-images.githubusercontent.com/2223916/84549974-2e5f9080-acbe-11ea-925a-92e2211ed9a2.png)

mazameli commented:
It's looking to me like this is the main scenario where things are fitting least nicely currently. I think I'd either rotate 90 degrees, or start hiding values.

![image](https://user-images.githubusercontent.com/2223916/84550677-14bf4880-acc0-11ea-8d48-b22d3f73b298.png)

To elaborate, it seems like Bars are effectively using the "Show all values" strategy even when "As many as can fit nicely" is selected. We're clearly gonna need to hide some values here:

![image](https://user-images.githubusercontent.com/2223916/84550996-10475f80-acc1-11ea-830c-7f3b48fcd2d1.png)

---

This scenario is of course going to look crazy almost no matter what, but I think we're actually handling it relatively well given how many series are on the screen:

![image](https://user-images.githubusercontent.com/2223916/84550853-9e6f1600-acc0-11ea-9b11-7b0b4f55e1a1.png)

And it's cool I can easily do this now:

![image](https://user-images.githubusercontent.com/2223916/84550884-b5ae0380-acc0-11ea-92dc-079b341cae86.png)

mazameli commented:
I think you implicitly noted this, Paul, but it appears that Area and Combo charts haven't yet had their global Display settings moved above the individual series settings.

mazameli commented:
Continuing to see values oddly displaying in center of bars when split y-axes are used with multiple metrics and a single grouping (and split y-axes are often used automatically by default):

![image](https://user-images.githubusercontent.com/2223916/84551475-a29c3300-acc2-11ea-8179-9ca5cc3cd73f.png)

paulrosenzweig committed:
fix issue where labels didn't turn on

paulrosenzweig committed:
rotating grouped bar labels

paulrosenzweig committed:
avoid multiple labels with unaggregated data, fix bug with stacked

paulrosenzweig committed:
extract onRenderValueLabels into another file

paulrosenzweig committed:
fix bug getting bar width

paulrosenzweig committed:
ensure correct settings tab order

paulrosenzweig commented:
@mazameli I implemented the 90º labels. They rotate at a certain bar thickness and hide completely as it gets thinner:
![image](https://user-images.githubusercontent.com/691495/84698312-98c23c00-af1d-11ea-880f-12312f3ab214.png)
![image](https://user-images.githubusercontent.com/691495/84698256-7fb98b00-af1d-11ea-9647-3324ac782cfe.png)
![image](https://user-images.githubusercontent.com/691495/84698231-74665f80-af1d-11ea-85c3-8b0621fef14c.png)

---

I'm unable to reproduce the issue with split y axes. Though, it's possible I accidentally fixed it before trying to repro. The question from your screenshot looked like order count and average product price by order month, right? Any other things I should try?

mazameli commented:
@paulrosenzweig Cool, I'm gonna play around with this now. I'll see if I can still repro that split axis bug.

mazameli commented:
Looks like Area and Combo charts still have their global settings below the per-series settings?

![image](https://user-images.githubusercontent.com/2223916/84712169-c2c72e80-af1c-11ea-8691-1dcb9aba8707.png)

![image](https://user-images.githubusercontent.com/2223916/84712188-d1154a80-af1c-11ea-9035-1d46aa638a82.png)

paulrosenzweig committed:
reorder graph data settings to the end

paulrosenzweig commented:
@mazameli whoops. Just pushed a commit to reorder those.

Separately, we should probably explicitly define the ordering somewhere. I introduced that for the radio button tabs in this PR. It would be trickier but probably make sense to do that for settings in each tab too.

mazameli commented:
I don't think this is working quite right for scatter plots, and I'm happy to have scatter plot support be a stretch goal, since the main goal is to get this working for multi-series line/area/bar/combo charts.

The main missing thing I think is that scatter plots need a way for you to choose which column to use for the label, which could be an option similar to the one that lets you pick which column to use for the bubble size. I.e., maybe we do it on the Data tab for scatter plots instead of the Display tab.

Reason being, with scatter plots, you're often plotting unaggregated data, with two numeric variables as the x and y axes, in which case you'd want to choose which of those columns, or some other column altogether, that you'd want to use for the labels.

Or, you're doing an aggregation like I have here, with the sum of walks and home runs broken out by each baseball team, and I'd like to be able to choose to display the _team name_ as the label for each dot:

![image](https://user-images.githubusercontent.com/2223916/84712868-8dbbdb80-af1e-11ea-9e19-9e667a215c1a.png)

Currently in the branch, it looks like trying to display values for unaggregated data with a scatter plot results in floating labels at the top of the chart:

![image](https://user-images.githubusercontent.com/2223916/84713313-8f39d380-af1f-11ea-8074-2285d2baebb8.png)

mazameli commented:
> I'm unable to reproduce the issue with split y axes.

Neither am I. Looks like you fixed it. :)

mazameli commented:
I think we should hide the individual series toggle for value labels when there's only one series like this. It's goofy as is:

![image](https://user-images.githubusercontent.com/2223916/84714536-d07fb280-af22-11ea-8032-7f50e96172b6.png)

mazameli commented:
Here's a funny case: I have a single metric grouped by two columns, resulting in ~30 series. I'm stacking the bars here, and it appears there's plenty of room to show values on top of each stack, but they only show up if I choose the `All` values setting.

`As many as can fit nicely`

![image](https://user-images.githubusercontent.com/2223916/84715084-65cf7680-af24-11ea-9835-301214b2e10e.png)

`All`

![image](https://user-images.githubusercontent.com/2223916/84715097-71bb3880-af24-11ea-857b-d01225503e4c.png)

mazameli commented:
I think the rotated labels as a strategy for tightly spaced bars is working pretty well, though I'm thinking maybe we want to go white rather than outlined for the numbers. Only problem would be cases where white-labeled instances are using really light colors, though that's inadvisable for charts in any case.

Thoughts, @kdoh?

![image](https://user-images.githubusercontent.com/2223916/84715416-3c631a80-af25-11ea-9eb2-cac4ade167e2.png)

mazameli commented:
I'm looking for a way to reliably repro this, but I'm sometimes seeing rotated labels float off center from their corresponding bars:

![image](https://user-images.githubusercontent.com/2223916/84715928-c2cc2c00-af26-11ea-9955-e3ffa5c3375f.png)

paulrosenzweig committed:
don't show series level setting if...

- there's one series
- there's no global setting for this chart
- the charts are stacked at all (including normalized)

paulrosenzweig commented:
@mazameli I had meant to punt on scatter plots, but the series-level control was accidentally included. I updated it to only appear when the the global setting is present.

paulrosenzweig committed:
fix bug with overly sparse stacked bar labels

kdoh commented:
Definitely a +1 on moving to white colored value labels in cases where we know the value will be contained inside of a bar.

paulrosenzweig committed:
fix misalignment

paulrosenzweig committed:
white labels inside bars

paulrosenzweig commented:
White labels have been pushed. The one place where it looks a bit funny is when short bars with labels above get mixed in:
![image](https://user-images.githubusercontent.com/691495/84933795-6d1d8e00-b0a4-11ea-8801-65666d3c23bb.png)

mazameli commented:
In dashboard fullscreen mode, looks like we're enlarging labels within the bars, but not outside of the bars:

![image](https://user-images.githubusercontent.com/2223916/84962232-39019780-b0bb-11ea-8892-659b836e3443.png)

mazameli commented:
I think in cases like this one, we're better off hiding the black labels:

![image](https://user-images.githubusercontent.com/2223916/84962852-a06c1700-b0bc-11ea-96bc-9940586191f2.png)

paulrosenzweig committed:
remove rotated labels above bars, match label size on dashboards

paulrosenzweig commented:
I removed the rotated black labels and updated the white label size on dashboards to match.

mazameli committed:
make labels 12px on questions

mazameli commented:
I pushed a commit to make labels on questions 12px to match dashboards and I think it looks pretty nice.

**Couple more things I noticed:**

With combo charts, if my x-axis is a binned number, I'm seeing the bars go past the right edge of the x-axis:

<img width="232" alt="Screen Shot 2020-06-18 at 9 48 09 AM" src="https://user-images.githubusercontent.com/2223916/85050251-a8729800-b14a-11ea-8cc2-5762357af6b4.png">

Can we do anything about cases like this where labels are on top of each other?

<img width="251" alt="Screen Shot 2020-06-18 at 9 48 12 AM" src="https://user-images.githubusercontent.com/2223916/85050340-ca6c1a80-b14a-11ea-8125-4ab4634e7b74.png">

mazameli commented:
@kdoh Do we still believe in showing value labels on bar charts by default? I'm starting to feel like they should be opt-in instead of opt-out. They're kind of an editorial choice I feel like I would tend to make once I'm _done_ exploring (i.e., arrived at a final result), and I think they might be distracting from seeing the trend/distribution of values when doing ad hoc exploration.

kdoh commented:
@mazameli Yeah after playing with this a bit I think I'm with you on having them be opt in vs on by default.

kdoh reviewed (commented):
Code looks pretty solid. (Thanks for the comments!). Thinking this can get the 'ol green checkmark after the last couple bits Maz mentioned get done.

paulrosenzweig committed:
default labels to off, fix misalignment on binned combos, more cluster avoidance

paulrosenzweig commented:
Chart values are now opt-in.

---

I tweaked the binned combo charts to be avoid running over the edge. It still doesn't quite match a normal histogram, but it's less obviously a bug.

![image](https://user-images.githubusercontent.com/691495/85079264-6110f980-b194-11ea-963e-86479887acf0.png)

--- 
To handle this overlap situation, I included bars with lines in the logic to drop clustered labels. It's still not perfect (e.g. grouped bars with lines can get a bit weird), but I think it's better.

![image](https://user-images.githubusercontent.com/691495/85080473-4429f580-b197-11ea-9b09-462a63da277b.png)

paulrosenzweig committed:
small fixes

paulrosenzweig committed:
remove tests around default chart values for bar charts
